polygeist/tools/cgeist/Test/Verification/str.c

```diff
@@ -10,11 +10,13 @@ int meta() {
 // CHECK:   llvm.mlir.global internal constant @str0("bar\00")
 // CHECK:   func @meta() -> i32 attributes {llvm.linkage = #llvm.linkage<external>} {
 // CHECK-NEXT:     %0 = llvm.mlir.addressof @str0 : !llvm.ptr<array<4 x i8>>
-// CHECK-NEXT:     %1 = llvm.getelementptr %0[0, 0] : (!llvm.ptr<array<4 x i8>>) -> !llvm.ptr<i8>
-// CHECK-NEXT:     %2 = call @foo(%1) : (!llvm.ptr<i8>) -> i32
+// CHECK-NEXT:     %1 = "polygeist.pointer2memref"(%0) : (!llvm.ptr<array<4 x i8>>) -> memref<?xi8>
+// CHECK-NEXT:     %2 = call @foo(%1) : (memref<?xi8>) -> i32
```

Comment:
Is this cast valid?

Reply:
I checked the LLVM IR produced by `cgeist` and `clang`, they are equivalent (other than the missing nsw on the add instruction and the use of i32 for the GEP indexes):

`cgeist LLVM IR`
```
cgeist -O0 a.cpp -S -emit-llvm -function='*' > cgeist.ll                                    
warning: Not emitting overflow-checked add
 ~/projects/intel-llvm/build  use_memref_for_i8ptr 9  cat cgeist.ll                                                                                                                                  
; ModuleID = 'a.cpp'
source_filename = "a.cpp"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@str1 = internal constant [11 x i8] c"int meta()\00"
@str0 = internal constant [4 x i8] c"bar\00"

declare i8* @malloc(i64)

declare void @free(i8*)

define i32 @_Z4metav() {
  %1 = call i32 @_Z3fooPKc(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @str0, i32 0, i32 0))
  %2 = call i32 @_Z3fooPKc(i8* getelementptr inbounds ([11 x i8], [11 x i8]* @str1, i32 0, i32 0))
  %3 = add i32 %1, %2
  ret i32 %3
}
```

`clang LLVM IR`
```
clang++  -O0 a.cpp -S -emit-llvm; cat a.ll                                                             
; ModuleID = 'a.cpp'
source_filename = "a.cpp"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@.str = private unnamed_addr constant [4 x i8] c"bar\00", align 1
@__PRETTY_FUNCTION__._Z4metav = private unnamed_addr constant [11 x i8] c"int meta()\00", align 1

; Function Attrs: noinline optnone uwtable
define dso_local i32 @_Z4metav() #0 {
  %1 = call i32 @_Z3fooPKc(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @.str, i64 0, i64 0))
  %2 = call i32 @_Z3fooPKc(i8* getelementptr inbounds ([11 x i8], [11 x i8]* @__PRETTY_FUNCTION__._Z4metav, i64 0, i64 0))
  %3 = add nsw i32 %1, %2
  ret i32 %3
}
...
```

Reply:
Also, this program runs correctly:
```
#include <stdio.h>

void foo(const char *p) { printf("%s\n", p); }
int main() {
  foo("bar");
  foo(__PRETTY_FUNCTION__);
  return 0;
}
```
```
cgeist -O0 a.cpp -o a.exe ; ./a.exe              
bar
int main()
```